src/parser_lyb.c

```diff
@@ -234,14 +235,39 @@ lyb_read_string(char **str, ly_bool with_length, struct lylyb_ctx *lybctx)
  * @return LY_ERR value.
  */
 static LY_ERR
-lyb_read_term(uint8_t **term_value, uint32_t *term_value_len, struct lylyb_ctx *lybctx)
+lyb_read_term(const struct lysc_node_leaf *term, uint8_t **term_value,
+        uint32_t *term_value_len, struct lylyb_ctx *lybctx)
 {
     uint32_t allocated_size;
+    int32_t lyb_data_len;
+    struct lysc_type_leafref *type_lf;
+    struct lysc_type_bits *type_bits;
+
+    assert(term && term_value && term_value_len && lybctx);
+
+    /*  Find out the size from @ref howtoDataLYB. */
+    switch (term->type->basetype) {
+    case LY_TYPE_LEAFREF:
+        /* Leafref itself is ignored, the target is loaded directly. */
+        type_lf = (struct lysc_type_leafref *)term->type;
+        lyb_data_len = type_lf->realtype->plugin->lyb_data_len;
+        break;
+    case LY_TYPE_BITS:
+        type_bits = (struct lysc_type_bits *)term->type;
+        lyb_data_len = lyplg_type_bits_bitmap_size(type_bits);
+        break;
+    default:
+        lyb_data_len = term->type->plugin->lyb_data_len;
```

Comment:
The idea was that either the `lyb_data_len` is fixed, in case you use it and do not read it and if it is not (including `bits`), you read it from the data, nothing else, no additional logic.